### ISSUE TITLE
GDScript leaking potentially when cannot load due to cyclic dependency

### ISSUE THREAD

**Issue by hbina** [godotengine/godot#32378]
**Godot version:**
b2569d5fa8e45173bbd8c03d9887e7635c7986b0

**OS/device including version:**
Linux Ubuntu 5.3.1-050301-generic

**Issue description:**
Godot is leaking memory (failure to cleanup).
It possibly happen when GDScript fails to load a class due to cyclic dependency or parse error.

```
SCRIPT ERROR: GDScript::reload: Parse Error: The class "Controls" couldn't be fully loaded (script error or cyclic dependency).
   At: res://Scripts/Game.gd:82.
ERROR: reload: Method/Function Failed, returning: ERR_PARSE_ERROR
   At: modules/gdscript/gdscript.cpp:576.
SCRIPT ERROR: GDScript::reload: Parse Error: The class "Controls" couldn't be fully loaded (script error or cyclic dependency).
   At: res://Scripts/Networking/Network.gd:235.
ERROR: reload: Method/Function Failed, returning: ERR_PARSE_ERROR
   At: modules/gdscript/gdscript.cpp:576.
SCRIPT ERROR: GDScript::reload: Parse Error: The class "Controls" couldn't be fully loaded (script error or cyclic dependency).
   At: res://Scripts/Game.gd:82.
ERROR: reload: Method/Function Failed, returning: ERR_PARSE_ERROR
   At: modules/gdscript/gdscript.cpp:576.
ERROR: start: Timer was not added to the SceneTree!
   At: scene/main/timer.cpp:111.
ERROR: start: Timer was not added to the SceneTree!
   At: scene/main/timer.cpp:111.
SCRIPT ERROR: GDScript::reload: Parse Error: The class "Controls" couldn't be fully loaded (script error or cyclic dependency).
   At: res://Scripts/Networking/Network.gd:236.
ERROR: reload: Method/Function Failed, returning: ERR_PARSE_ERROR
   At: modules/gdscript/gdscript.cpp:576.
ERROR: ~List: Condition ' _first != __null ' is true.
   At: ./core/self_list.h:111.
ERROR: ~List: Condition ' _first != __null ' is true.
   At: ./core/self_list.h:111.
WARNING: cleanup: ObjectDB Instances still exist!
   At: core/object.cpp:2069.
ERROR: clear: Resources Still in use at Exit!
   At: core/resource.cpp:443.
```

**Steps to reproduce:**

1. Clone KoBeWi's https://github.com/KoBeWi/The-Soulhunter
2. Run the game at least once (there will be errors)
3. Exit

**Minimal reproduction project:**
https://github.com/KoBeWi/The-Soulhunter

**Comment by hbina** [comment]
Update: It will leak simply by opening and closing the project...Will try to investigate...

**Comment by clayjohn** [comment]
Can you still reproduce this bug in [Godot 3.2.2](https://godotengine.org/article/maintenance-release-godot-3-2-2) or any later release?

This may also be something that isn't being fixed until Godot 4.0 releases with the new improvements to GDScript. I remember vnen talking about improvements to cyclic dependencies.

**Comment by akien-mga** [comment]
Duplicate of #16959.
